### ISSUE TITLE
[Feature Request] Custom Actions

### ISSUE THREAD

**Issue by DellanX** [dipscope/EntityStore.TS#21]
**Is your feature request related to a problem? Please describe.**

I'm hoping for a more streamlined process to execute some custom actions on my API server, which results in normal JSON:API responses. For example, I have an endpoint on my server to process some status information from an object. `/{type}/{id}/c-actions/status`, this object returns a JSON:API payload of the type `statuses`. (I use c-actions in my route, that being said, JSON:API doesn't have an official custom action structure yet)

Unfortunately, due to a complex issue with the server, I am not able to provide 'status' as a relationship.
(Server is Laravel, and there is some issues when not using their builtin ORM)

**Describe the solution you'd like**

In an ideal world, I could execute some code with the following structure to execute some requests on the server.
```ts
const status = entityStore.createEntitySet(User).executeCustomFind(user, '/c-actions', Status);
status.category = 'busy';
entityStore.createEntitySet(User).executeCustomUpdate(user, '/status', status);
```
Or, to steal a page from the `JsonApiEntityProvider.TS` package:
```ts
const customActionProvider = entityStore.createEntitySet(User).createCustomActionProvider(user, '/c-actions');
const status = customActionProvider.find(Status, '/status');
status.category = 'busy';
customActionProvider.update(status, '/status');
```

**Describe alternatives you've considered**

For my use case, I am probably going to just use the Fetch API to make the requests, then use TypeManager to serialize/deserialize.
That being said, I have some JSON:RPC commands I'd like to bring over to JSON:API, and this would be the perfect area.

**Additional context**

As always, I am willing to help if given direction.
This library is amazing!

**Comment by dpimonov** [comment]
Hi [DellanX](https://github.com/DellanX),

Long time no see! Thanks for the issue! I think it fits more to the `JsonApiEntityProvider` but we can keep it here. :)

About your case - I think more about simply providing one optional parameter on relationship links which allows override default `relationships` path to something else. Currently we have:

```typescript
const userMessages = jsonApiEntityProvider.createJsonApiToManyRelationship(userSet, user, u => u.messages);
```

This will create a following link:

`users/id/relationships/messages`

If we provide additional argument to override the relationships path which is currently hardcoded here:

![image](https://github.com/dipscope/EntityStore.TS/assets/15626139/fa8327a1-2aa9-4fb8-af4a-b026d4fdf5ac)

We can archive behaviour you want if I've got it correctly. So relationship creation will become:

```typescript
const userMessages = jsonApiEntityProvider.createJsonApiToManyRelationship(userSet, user, u => u.messages, 'c-actions');
```

Last argument by default is `relationships` so it should stay optional but one can override it with something else to create custom relationship links like in your case:

`users/id/c-actions/messages`

So instead of hardcoding `relationships` string in `JsonApiAdapter` method - we pass it when creating a relationship. The messages property can also be overriden already by using `alias` option as well as excluded from serialization if required. So the whole relationship path became configurable. 

What do you think? :)

**Comment by DellanX** [comment]
Oh wow! I think that's a great idea! It'd most definitely the problem I care about.

I do wonder about the `update` use case. For example, if I send up a POST/PUT request with Type A, I might be recieving type B. (Friend Request Approval converts a `FriendRequest` into a `Friendship`).

That being said, this special use case isn't that important to me, as I can keep my current process without headache.
* I use JSON:RPC to send the approval, then on success, I use JSON:API to fetch the new `Friendship`

Are there any tasks you'd like me to help you with?

**Comment by dpimonov** [comment]
Sorry for long responses. I have a tight schedule right now but can implement it next week. The larger part are tests as usual. You can help me to provide there your use cases if they are possible to reproduce with framework i use for tests. If it is urgent - you can implement a feature as well. It is simply about passing one parameter from relationship link to adapter.

**Comment by dpimonov** [comment]
Hey [DellanX](https://github.com/DellanX)[DellanX](https://github.com/DellanX), I released a version of json api entity provider with additional option to override relationships path. You can give it a try. If it works - you can write one or two additional tests for related relationship provider if you have time.

**Comment by DellanX** [comment]
I'm working on some tests, now that I can invoke a mock fetch method, but I'm not seeing any behavior changes when specifying a path:

Both `find` calls below query the same route (under relationships)
```ts

    const userCompany = jsonApiEntityProvider.createJsonApiToOneRelationship(userSet, user, u => u.company);
    const userCompanyCustom = jsonApiEntityProvider.createJsonApiToOneRelationship(userSet, user, u => u.company, 'custom-action');


        const data = await userCompany.find();
        const data = await userCompanyCustom.find();
```

I'll create a second PR on top of my fetchInjector PR to add the test so that we can tinker with it.

**Comment by dpimonov** [comment]
Solved in v2.4.0 of `JsonApiEntityProvider`.
